{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_chromosome_sequence(fna_file, chromosome_id):\n",
    "    # Parse the .fna file using SeqIO\n",
    "    for record in SeqIO.parse(fna_file, \"fasta\"):\n",
    "        # Check if the chromosome ID is in the record description\n",
    "        if chromosome_id in record.description:\n",
    "            print(f\"Found chromosome: {record.description}\")\n",
    "            return str(record.seq)\n",
    "    # Return None if the chromosome is not found\n",
    "    return None\n",
    "\n",
    "def make_windows(w_plus_k, sequence):\n",
    "    # make a dictionary of windows of size w+k to count\n",
    "    windows = dict()\n",
    "    for i in range(len(sequence)-w_plus_k+1):\n",
    "        window = sequence[i:i+w_plus_k]\n",
    "        if window in windows:\n",
    "            windows[window] += 1\n",
    "        else:\n",
    "            windows[window] = 1\n",
    "\n",
    "    # save to a .csv file\n",
    "    df = pd.DataFrame(windows.items(), columns=['window', 'count'])\n",
    "    df.to_csv(f'windows_{w_plus_k}.csv', index=False)\n",
    "\n",
    "    # print the number of windows\n",
    "    print(f\"Number of windows of size {w_plus_k}: {len(windows)}\")\n",
    "\n",
    "    # print the average count of windows\n",
    "    print(f\"Average count of windows of size {w_plus_k}: {df['count'].mean()}\")\n",
    "\n",
    "\n",
    "\n",
    "def make_naive_windows(w_plus_k, sequence, name):\n",
    "    # make a list of windows of size w+k \n",
    "    windows = []\n",
    "    #print(type(w_plus_k))\n",
    "    for i in range(len(sequence)-w_plus_k+1):\n",
    "        window = sequence[i:i+w_plus_k]\n",
    "        windows.append(window)\n",
    "\n",
    "    # save to a .csv file\n",
    "    df = pd.DataFrame(windows, columns=['window'])\n",
    "    subfolder = f\"sequences_{name}\"\n",
    "    df.to_csv(f'{subfolder}/naive_windows_{w_plus_k}.csv', index=False)\n",
    "\n",
    "\n",
    "\n",
    "def dna_to_binary_string(dna):\n",
    "    binary_mapping = {'A': '00', 'C': '01', 'G': '10', 'T': '11'}\n",
    "    return ''.join([binary_mapping[nuc] for nuc in dna])\n",
    "\n",
    "\n",
    "def delete_naive_window(size, name):\n",
    "    subfolder = f\"sequences_{name}\"\n",
    "    os.remove(f'{subfolder}/naive_windows_{size}.csv')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def naive_window_to_odd_even(size, name):\n",
    "    subfolder = f\"sequences_{name}\"\n",
    "    df = pd.read_csv(f'{subfolder}/naive_windows_{size}.csv')\n",
    "\n",
    "    windows = df['window'].to_list()\n",
    "    \n",
    "    results = []\n",
    "    for window in windows:\n",
    "        binary_string = dna_to_binary_string(window)\n",
    "        \n",
    "        # Step 3: Separate into odd and even positions\n",
    "        odd_positions = binary_string[0::2]\n",
    "        even_positions = binary_string[1::2]\n",
    "        \n",
    "        # Step 4: Store as integers\n",
    "        odd_number = int(''.join(odd_positions), 2)\n",
    "        even_number = int(''.join(even_positions), 2)\n",
    "        \n",
    "        results.append({'OddNumber': odd_number, 'EvenNumber': even_number})\n",
    "\n",
    "    # Convert the results to a DataFrame for easy viewing\n",
    "    result_df = pd.DataFrame(results)\n",
    "\n",
    "    # Save the results to a CSV file\n",
    "    result_df.to_csv(f'{subfolder}/contexts_{size}_odd_even.csv', index=False)\n",
    "\n",
    "\n",
    "\n",
    "def save_as_fasta(sequence, filename):\n",
    "    with open(filename, 'w') as f:\n",
    "        f.write(f'<\\n{sequence}')\n",
    "\n",
    "def save_as_seq(sequence, filename):\n",
    "    with open(filename, 'w') as f:\n",
    "        f.write(f'{sequence}\\n')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "FILE_PATH = 'GCA_009914755.4_T2T-CHM13v2.0_genomic.fna' # Change this to the path of the .fna file\n",
    "chromosome_id = \"chromosome X\"   # Here we extract the sequence of chromosome X out of the .fna file\n",
    "\n",
    "sizes = [1000000] # Change this if you want a smaller sequence size (useful mainly for testing)\n",
    "names = ['1M'] # Change this to the name of the sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found chromosome: CP068255.2 Homo sapiens isolate CHM13 chromosome X\n"
     ]
    }
   ],
   "source": [
    "# get the sequence of the chromosome - note that this can be done in other ways if your data is in a different format,\n",
    "# The only important thing is that at this stage the variable \"sequence\" contains a string representation of the sequence in uppercase ACGT\n",
    "sequence = read_chromosome_sequence(FILE_PATH, chromosome_id)\n",
    "sequence = sequence.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating 1M sequences\n",
      "Generating 1M sequences with context size 2\n",
      "Generating 1M sequences with context size 3\n",
      "Generating 1M sequences with context size 4\n",
      "Generating 1M sequences with context size 5\n",
      "Generating 1M sequences with context size 6\n",
      "Generating 1M sequences with context size 7\n",
      "Generating 1M sequences with context size 8\n",
      "Generating 1M sequences with context size 9\n",
      "Generating 1M sequences with context size 10\n",
      "Generating 1M sequences with context size 11\n",
      "Generating 1M sequences with context size 12\n",
      "Generating 1M sequences with context size 13\n",
      "Generating 1M sequences with context size 14\n",
      "Generating 1M sequences with context size 15\n",
      "Generating 1M sequences with context size 16\n",
      "Generating 1M sequences with context size 17\n",
      "Generating 1M sequences with context size 18\n",
      "Generating 1M sequences with context size 19\n",
      "Generating 1M sequences with context size 20\n",
      "Generating 1M sequences with context size 21\n",
      "Generating 1M sequences with context size 22\n",
      "Generating 1M sequences with context size 23\n",
      "Generating 1M sequences with context size 24\n",
      "Generating 1M sequences with context size 25\n",
      "Generating 1M sequences with context size 26\n",
      "Generating 1M sequences with context size 27\n",
      "Generating 1M sequences with context size 28\n",
      "Generating 1M sequences with context size 29\n",
      "Generating 1M sequences with context size 30\n",
      "Generating 1M sequences with context size 31\n",
      "Generating 1M sequences with context size 32\n",
      "Generating 1M sequences with context size 33\n",
      "Generating 1M sequences with context size 34\n",
      "Generating 1M sequences with context size 35\n",
      "Generating 1M sequences with context size 36\n",
      "Generating 1M sequences with context size 37\n",
      "Generating 1M sequences with context size 38\n",
      "Generating 1M sequences with context size 39\n"
     ]
    }
   ],
   "source": [
    "for size, name in zip(sizes,names):\n",
    "    print(f\"Generating {name} sequences\")\n",
    "    shorter_sequence = sequence[:size] # Can remove this and keep shorter_sequence = sequence if not interested in a subset of the sequence\n",
    "\n",
    "    # create a subfolder for the size\n",
    "    subfolder = f\"sequences_{name}\"\n",
    "    os.makedirs(subfolder, exist_ok=True)\n",
    "    \n",
    "    # save it to a .txt file\n",
    "    with open(f\"{subfolder}/sequence_{name}.txt\", 'w') as f:\n",
    "        f.write(shorter_sequence)\n",
    "\n",
    "    # make contexts of various w+k sizes.\n",
    "    # For instance if we are interested in k=4 and w =10, we need to just calculate this for \"window_size\"=14 and not the entire range\n",
    "    for window_size in range (2, 40):\n",
    "        print(f\"Generating {name} sequences with context size {window_size}\")\n",
    "        make_naive_windows(window_size, shorter_sequence, name)\n",
    "        naive_window_to_odd_even(window_size, name)\n",
    "        delete_naive_window(window_size, name)\n",
    "    # This was done in order to compare to other methods, but it is not necessary for generating low particular density orders.\n",
    "    save_as_fasta(shorter_sequence, f'{subfolder}/sequence_{name}.fasta')\n",
    "    save_as_seq(shorter_sequence, f'{subfolder}/sequence_{name}.seq')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
